File: 22. Generate Parentheses-1.cpp
class Solution {
public:
  vector<string> res;
  vector<string> generateParenthesis(int n) {
    string s;
    backtrack(s, 0, 0, n);

    return res;
  }

  void backtrack(string &s, int open, int close, int n) {
    if (s.size() == n * 2) {
      res.push_back(s);
      return;
    }

    if (open < n) {
      s.push_back('(');
      backtrack(s, open + 1, close, n);
      s.pop_back();
    }

    if (close < open) {
      s.push_back(')');
      backtrack(s, open, close + 1, n);
      s.pop_back();
    }
  }
};

File: 22. Generate Parentheses-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        m = n * 2
        res = []
        path = [""] * m

        def dfs(i, open):
            if i == m:
                res.append("".join(path))
                return
            if open < n:
                path[i] = "("
                dfs(i + 1, open + 1)
            if i - open < open:
                path[i] = ")"
                dfs(i + 1, open)

        dfs(0, 0)
        return res
